### ISSUE TITLE
AWSCognitoIdentityProvider: Sign in using alias (email)

### ISSUE THREAD

**Issue by stephencroberts** [aws/aws-sdk-ios#373]
I'm having trouble with the iOS SDK signing a user in using a verified alias. I set up a browser demo with the JS SDK and was able to authenticate the user with no problem with both the username and the email address, but the iOS SDK only lets me authenticate with the username. When I use the email address, it gives me a `Incorrect username or password.` response. The following line of code performs the sign-in:

```swift
pool.getUser().getSession(usernameTextField.text!, password: passwordTextField.text!, validationData: nil, scopes: nil)
```

Is this a bug in the iOS sdk or am I missing something?

**Comment by behrooziAWS** [comment]
I'll see if I can repro this and get back to you.

**Comment by behrooziAWS** [comment]
Yes, it's a bug in the iOS SDK.  We will patch this in the next release.

If you need to get this working in the interim, replace [this code in AWSCognitoIdentityUser.m](https://github.com/aws/aws-sdk-ios/blob/master/AWSCognitoIdentityProvider/AWSCognitoIdentityUser.m#L288-L297) with:

```objc
        srpHelper.clientState.userName = authDetails.username;
        NSData *hash = [srpHelper completeAuthentication:serverState];
        AWSCognitoIdentityProviderAuthenticateRequest *authInput = [AWSCognitoIdentityProviderAuthenticateRequest new];
        authInput.clientId = self.pool.userPoolConfiguration.clientId;
        authInput.username = authDetails.username;
        authInput.timestamp = srpHelper.clientState.timestamp;
        authInput.passwordClaim = [[AWSCognitoIdentityProviderPasswordClaimType alloc] init];
        authInput.passwordClaim.secretBlock = authDetails.secretBlock;
        authInput.passwordClaim.signature = hash;
        authInput.secretHash = [self.pool calculateSecretHash:authDetails.username];
        return [self.pool.client authenticate:authInput];
```

**Comment by stephencroberts** [comment]
Excellent work, thank you! I'm really excited about this feature from AWS.

**Comment by behrooziAWS** [comment]
This has been fixed in 2.4.2.  Let me know if you still encounter issues.
